business/authentication/openid_auth_controller.go

```diff
@@ -0,0 +1,1444 @@
+		p.Error = &AuthenticationFailureError{
+			Reason: "the received id_token from the OpenId provider has missing the required 'exp' claim",
+		}
+		p.ShouldTerminateSession = true
+		return p
+	} else {
+		// If the expiration date is present on the claim, we use that
+		expiresInNumber, err := parseTimeClaim(expClaim)
+		if err != nil {
+			p.Error = &AuthenticationFailureError{
+				Reason: "token exp claim is present, but invalid",
+				Detail: err,
+			}
+			p.ShouldTerminateSession = true
+			return p
+		}
+
+		p.ExpiresOn = time.Unix(expiresInNumber, 0)
+	}
+
+	// Extract the name of the user from the id_token. The "subject" is passed to the front-end to be displayed.
+	p.Subject = "OpenId User" // Set a default value
+	if userClaim, ok := claims[config.Get().Auth.OpenId.UsernameClaim]; ok && len(userClaim.(string)) > 0 {
+		p.Subject = userClaim.(string)
+	}
+
+	return p
+}
+
+// validateOpenIdNonceCode checks that the nonce hash that is present in the id_token is the right
+// hash, given the nonce code present in the http cookie.
+//
+// This is the replay attack mitigation.
+func (p *openidFlowHelper) validateOpenIdNonceCode() *openidFlowHelper {
+	// Do nothing if there was an error in previous flow steps.
+	if p.Error != nil {
+		return p
+	}
+
+	// Parse the received id_token from the IdP and check nonce code
+	nonceHashHex := fmt.Sprintf("%x", p.NonceHash)
+	if nonceClaim, ok := p.IdTokenPayload["nonce"]; !ok || nonceHashHex != nonceClaim.(string) {
+		p.Error = &AuthenticationFailureError{
+			HttpStatus: http.StatusForbidden,
+			Reason:     "OpenId token rejected: nonce code mismatch",
+		}
+	}
+	return p
+}
+
+// validateOpenIdState verifies that the "state" parameter passed during the callback to Kiali
+// has the expected value, given the value of the nonce cookie and Kiali's signing key.
+//
+// This is the CSRF attack mitigation.
+func (p *openidFlowHelper) validateOpenIdState() *openidFlowHelper {
+	// Do nothing if there was an error in previous flow steps.
+	if p.Error != nil {
+		return p
+	}
+
+	state := p.State
+
+	separator := strings.LastIndexByte(state, '-')
+	if separator != -1 {
+		csrfToken, timestamp := state[:separator], state[separator+1:]
+		csrfHash := sha256.Sum224([]byte(fmt.Sprintf("%s+%s+%s", p.Nonce, timestamp, config.GetSigningKey())))
+
+		if fmt.Sprintf("%x", csrfHash) != csrfToken {
+			p.Error = &AuthenticationFailureError{
+				HttpStatus: http.StatusForbidden,
+				Reason:     "Request rejected: CSRF mitigation",
+			}
+		}
+	} else {
+		p.Error = &AuthenticationFailureError{
+			HttpStatus: http.StatusForbidden,
+			Reason:     "Request rejected: State parameter is invalid",
+		}
+	}
+
+	return p
+}
+
+// requestOpenIdToken makes a request to the OpenId server to exchange the received code (of the
+// authorization code flow) with a proper identity token (id_token) and an access_token (if applicable).
+func (p *openidFlowHelper) requestOpenIdToken(redirect_uri string) *openidFlowHelper {
```

Comment:
Would the [stdlib oauth2 package](https://pkg.go.dev/golang.org/x/oauth2) help in doing some of this flow such as the token exchange?

Reply:
Yes!
I remember I was considering using that lib before the very first version of all this code. To be honest, I don't clearly remember why I didn't use it. What I remember is that the oauth2 package only deals with "OAuth", but not OpenID - although they are very similar how they work.

I also considered using [coreos/go-oidc](https://github.com/coreos/go-oidc), and the reason this one was not used is that it only supports the "authorization code flow" while back then Kiali only supported the "implicit flow". Nowdays, Kiali support both "implicit flow" and "authorization code flow". So, this lib still is still short.

I know all this doesn't really answers your question. I can't remember the reason :-(

Reply:
No worries I was just curious to know. Thanks for the info.